### PR TITLE
Added check to verify xla device is TPU

### LINKED ISSUE #3104: TPU available: true when there are no TPUs
## 🐛 Bug
I am using a DGX machine (and so, no TPUs), but on initiating `Trainer`, it logs `TPU available: True`. This ends up returning `Missing XLA configuration` when I run my script.

#### Code sample
Simply running the following lines on my machine:
```python
>> trainer = pl.Trainer(gpus=[0])                                                                                                                 
GPU available: True, used: True
TPU available: True, using: 0 TPU cores
```

### Expected behavior
```python
>> trainer = pl.Trainer(gpus=[0])                                                                                                                 
GPU available: True, used: True
TPU available: False, using: 0 TPU cores
```

### Environment
```
* CUDA:
        - GPU:
                - Tesla V100-SXM2-32GB
        - available:         True
        - version:           10.2
* Packages:
        - numpy:             1.18.2
        - pyTorch_debug:     False
        - pyTorch_version:   1.6.0
        - pytorch-lightning: 0.9.0
        - tensorboard:       2.2.0
        - tqdm:              4.45.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - 
        - processor:         x86_64
        - python:            3.6.9
        - version:           #168-Ubuntu SMP Wed Jan 16 21:00:45 UTC 2019
```

### PR DESCRIPTION
**Pull request by lezwon**
## What does this PR do?

Fixes #3104

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/models/test_tpu.py [hunk 1]
```diff
@@ -14,15 +15,18 @@
...
-    # TPU_AVAILABLE = device_type == 'TPU'
+    def tpu_device_exists(q):
+        device = torch_xla.core.xla_model.xla_device()
+        device_type = torch_xla.core.xla_model.xla_device_hw(device)
+        q.put(device_type == 'TPU')
+
+    queue = Queue()
+    p = Process(target=tpu_device_exists, args=(queue, ))
+    p.start()
+    p.join()
```
Comment (Borda) [inline tests/models/test_tpu.py:26]: can we make it as a function?

Comment (lezwon) [reply tests/models/test_tpu.py]: yep. working on it. :]

**Comment by lezwon** [comment]
@Borda @justusschock @awaelchli I need some help fixing the windows tests. Has to do something with pickling.

**Comment by awaelchli** [comment]
@lezwon I think this is a common subprocess thing, when you launch it, it needs to import everything first before it runs it, and so we have to guard the call from the import, i.e., you would have to add `if __name__ == "__main__"` around the TPU_AVAILABLE constant, and also remove the decorator and apply it directly to the call. This will work:

``` python
def tpu_device_exists():
    if xm is not None:
        device = xm.xla_device()
        device_type = fetch_xla_device_type(device)
        return device_type == "TPU"
    else:
        return False


if __name__ == "__main__":
    TPU_AVAILABLE = pl_multi_process(tpu_device_exists)()
    print(TPU_AVAILABLE)
```
but it's obviously not how we want it. 

My  suggestions is to move towards a functional check, like I propose here #2877. Then, we shouldn't get the pickle error since the value is only computed at runtime.

**Comment by edenlightning** [comment]
hwy @lezwon mind taking a look at this again? we would love to have this issue resolved

**Comment by lezwon** [comment]
@edenafek will continue on it by the weekend. 
Might need some help from @awaelchli on this.

**Comment by Borda** [comment]
@lezwon mind resolve conflicts so we know what is the actual state...

**Comment by awaelchli** [comment]
@lezwon do you still need help with this? is there a problem with windows?

**Comment by lezwon** [comment]
> @lezwon do you still need help with this? is there a problem with windows?

hey, I got it working with a minor workaround. Not sure about the cause of that issue though :)

**Comment by awaelchli** [comment]
is it ready for review @lezwon

**Comment by lezwon** [comment]
> is it ready for review @lezwon ?

yep. it is

File: pytorch_lightning/utilities/xla_device_utils.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+from multiprocessing import Process, Queue
+
+TORCHXLA_AVAILABLE = importlib.util.find_spec("torch_xla") is not None
+if TORCHXLA_AVAILABLE:
+    import torch_xla.core.xla_model as xm
+else:
+    xm = None
+
+
+def inner_f(queue, func, **kwargs):
```
Comment (awaelchli) [inline pytorch_lightning/utilities/xla_device_utils.py:12]:
```suggestion
def inner_f(queue, func, **kwargs):  # pragma: no cover
```
I am not 100% sure, but we may want to ignore the codecoverage on some of these multiprocessing parts.
For example, inner_f is launched with a new process, I'm pretty sure codecoverage cannot reach into it.

File: pytorch_lightning/utilities/xla_device_utils.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    try:
+        queue.put(func(**kwargs))
+    except Exception as _e:
+        import traceback
+        traceback.print_exc()
+        queue.put(None)
+
+
+def pl_multi_process(func):
```
Comment (awaelchli) [inline pytorch_lightning/utilities/xla_device_utils.py:21]: there is a similar, if not identical function in `tests/base/develop_utils.py`
do we need both? I cannot see an obvious difference besides the inner_f being defined outside.

Comment (lezwon) [reply pytorch_lightning/utilities/xla_device_utils.py:21]: I basically used the other function itself as reference. the one in `develop_utils` is just for tests right?

File: tests/utilities/test_xla_device_utils.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+@pytest.mark.skipif(not XLA_AVAILABLE, reason="test requires torch_xla to be installed")
+def test_tpu_device_does_not_exist():
+    """Check tpu_device_exists returns True when TPU is available"""
+    assert xdu.tpu_device_exists() is True
+
+
+@pytest.mark.skipif(not XLA_AVAILABLE, reason="test requires torch_xla to be installed")
+@pl_multi_process_test
+def test_tpu_device_does_not_exist():
```
[the remark this answers is not in the document]
Comment (lezwon) [reply tests/utilities/test_xla_device_utils.py:27]: oops. accidental error. Will fix it :)

**Comment by Borda** [comment]
@Lezwon mind rebase, these tests shall be fixed on master...

**Comment by lezwon** [comment]
@Borda done 👍 Some tests do occasionally fail though

File: tests/utilities/test_xla_device_utils.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import pytest
+
+from pytorch_lightning.utilities.xla_device_utils import XLADeviceUtils as xdu
+from tests.base.develop_utils import pl_multi_process_test
+
+try:
+    import torch_xla.core.xla_model as xm
+    XLA_AVAILABLE = True
+except ImportError as e:
```
Comment (Borda) [inline tests/utilities/test_xla_device_utils.py:9]:
```suggestion
except ImportError:
```

**Review by Borda — APPROVED** [review]
LGTM :]

File: pytorch_lightning/accelerators/tpu_backend.py [hunk 6]
```diff
@@ -169,7 +168,7 @@ def to_device(self, batch):
         See Also:
             - :func:`~pytorch_lightning.utilities.apply_func.move_data_to_device`
         """
-        if not XLA_AVAILABLE:
+        if not TPU_AVAILABLE:
             raise MisconfigurationException(
                 'Requested to transfer batch to TPU but XLA is not available.'
                 ' Are you sure this machine has TPUs?'
```
Comment (rohitgr7) [inline pytorch_lightning/accelerators/tpu_backend.py:174]: should this check be added in Trainer `__init__` itself for both XLA and TPU if `tpu_cores` is requested?? There might be cases when either of them is missing. What do you think??

File: tests/utilities/test_xla_device_utils.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+@pytest.mark.skipif(XLA_AVAILABLE, reason="test requires torch_xla to be absent")
+def test_tpu_device_absence():
+    """Check tpu_device_exists returns None when torch_xla is not available"""
+    assert xdu.tpu_device_exists() is None
+
+
+@pytest.mark.skipif(not XLA_AVAILABLE, reason="test requires torch_xla to be installed")
+def test_tpu_device_presence():
+    """Check tpu_device_exists returns True when TPU is available"""
+    assert xdu.tpu_device_exists() is True
```
Comment (rohitgr7) [inline tests/utilities/test_xla_device_utils.py:22]:
```suggestion
    assert xdu.tpu_device_exists()
```

File: pytorch_lightning/accelerators/tpu_backend.py (hunk 6 above)
Comment (lezwon) [reply pytorch_lightning/accelerators/tpu_backend.py:174]: The check for `gpus` is present in `__init__` yes. I guess we could refactor the code here and move the TPU check to trainer `__init__` too. A separate PR for that maybe? :)

Comment (Borda) [reply pytorch_lightning/accelerators/tpu_backend.py:174]: good, then, pls add TODO comment so we do not forget...

File: pytorch_lightning/utilities/xla_device_utils.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
+import functools
+import importlib
+from multiprocessing import Process, Queue
+
+TORCHXLA_AVAILABLE = importlib.util.find_spec("torch_xla") is not None
```
Comment (justusschock) [inline pytorch_lightning/utilities/xla_device_utils.py:5]: What is the benefit of this over catching the try and catch the import error?

File: pytorch_lightning/utilities/xla_device_utils.py (hunk 3 above)
Comment (justusschock) [reply pytorch_lightning/utilities/xla_device_utils.py:21]: yeah, but couldn't you delete the one from tests and then import from here?

**Review by justusschock — COMMENTED** [review]
Could you also add typing and docstrings for your TPU utils? :)

**Comment by Borda** [comment]
@lezwon I have rebased it, mind check if it is correct... :]

**Review by awaelchli — COMMENTED** [review]
lgtm
please check the comment by Justus:
https://github.com/PyTorchLightning/pytorch-lightning/pull/3274/files#r494993401

File: pytorch_lightning/utilities/xla_device_utils.py (hunk 3 above)
Comment (lezwon) [reply pytorch_lightning/utilities/xla_device_utils.py:21]: The `pl_multi_process_test` varies a bit compared to this function. It has an `assert` statement within and returns either `1` or `-1` for the test. This one is meant to return the device type on `None`.

File: pytorch_lightning/utilities/xla_device_utils.py (hunk 8 above)
Comment (lezwon) [reply pytorch_lightning/utilities/xla_device_utils.py:5]: This is a cleaner way to check if `torch_xla` exists. An advantage over try/catch would be that the `ImportError` could be raised from within the module itself, and try/catch would interpret that as `torch_xla` is not available. This method specifically checks for `torch_xla`.

**Comment by lezwon** [comment]
> @lezwon I have rebased it, mind check if it is correct... :]

Thank you :) It looks good to me 👍

File: tests/utilities/test_xla_device_utils.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+
+@pytest.mark.skipif(not XLA_AVAILABLE, reason="test requires torch_xla to be installed")
+@pl_multi_process_test
+def test_xla_device_is_a_tpu():
+    """Check that the XLA device is a TPU"""
+    device = xm.xla_device()
+    device_type = xm.xla_device_hw(device)
+    return device_type == "TPU"
```
Comment (rohitgr7) [inline tests/utilities/test_xla_device_utils.py:31]: calling `xm.xla_device()` explicitly won't break the TPU tests right if we add more tests later after this one?

File: tests/utilities/test_xla_device_utils.py (hunk 5 above)
Comment (rohitgr7) [inline tests/utilities/test_xla_device_utils.py:9]: use `TPU_AVAILABLE` here too instead of `XLA_AVAILABLE`?

File: tests/utilities/test_xla_device_utils.py (hunk 9 above)
Comment (lezwon) [reply tests/utilities/test_xla_device_utils.py:31]: Nope. It should be safe as the test is being run inside `pl_multi_process_test` decorator.

File: tests/utilities/test_xla_device_utils.py (hunk 5 above)
Comment (lezwon) [reply tests/utilities/test_xla_device_utils.py:9]: The purpose of the tests will be missed if I do that. The `XLA_AVAILABLE` variable makes sure that `torch_xla` module is present on the system. The tests determine if the XLA device is a TPU.

File: tests/utilities/test_xla_device_utils.py (hunk 9 above)
Comment (rohitgr7) [reply tests/utilities/test_xla_device_utils.py:31]: Oh ok didn't see that. Nice!

**Review by rohitgr7 — APPROVED** [review]
LGTM

File: pytorch_lightning/utilities/xla_device_utils.py (hunk 8 above)
Comment (Borda) [reply pytorch_lightning/utilities/xla_device_utils.py:7]: this is the way @awaelchli proposes...

**Review by SkafteNicki — APPROVED** [review]
